File: internal/session/session.go
// Package session persists pending per-user download sessions in SQLite.
package session

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"

	_ "modernc.org/sqlite" // pure-Go SQLite driver (no cgo)
)

// Data is the pending session for a user: the URL they sent and its metadata.
type Data struct {
	URL   string `json:"url"`
	Title string `json:"title"`
	ID    string `json:"id"`
}

// Store is a SQLite-backed session store.
type Store struct {
	db *sql.DB
}

// Open opens (creating if needed) the store at "<base>.sqlite3".
func Open(base string) (*Store, error) {
	path := base + ".sqlite3"
	if parent := filepath.Dir(path); parent != "" {
		if err := os.MkdirAll(parent, 0o755); err != nil {
			return nil, err
		}
	}
	db, err := sql.Open("sqlite", path)
	if err != nil {
		return nil, err
	}
	if _, err := db.Exec(
		"CREATE TABLE IF NOT EXISTS sessions (user_id INTEGER PRIMARY KEY, data TEXT NOT NULL)",
	); err != nil {
		_ = db.Close()
		return nil, err
	}
	return &Store{db: db}, nil
}

// Close releases the underlying database.
func (s *Store) Close() error { return s.db.Close() }

// Save upserts the session for userID.
func (s *Store) Save(userID int64, d Data) error {
	blob, err := json.Marshal(d)
	if err != nil {
		return err
	}
	_, err = s.db.Exec(
		"INSERT INTO sessions (user_id, data) VALUES (?, ?) "+
			"ON CONFLICT(user_id) DO UPDATE SET data = excluded.data",
		userID, string(blob),
	)
	return err
}

// Load returns the session for userID, or (nil, nil) if none exists.
func (s *Store) Load(userID int64) (*Data, error) {
	var blob string
	err := s.db.QueryRow("SELECT data FROM sessions WHERE user_id = ?", userID).Scan(&blob)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	var d Data
	if err := json.Unmarshal([]byte(blob), &d); err != nil {
		return nil, fmt.Errorf("decode session %d: %w", userID, err)
	}
	return &d, nil
}

// Delete removes the session for userID (no error if absent).
func (s *Store) Delete(userID int64) error {
	_, err := s.db.Exec("DELETE FROM sessions WHERE user_id = ?", userID)
	return err
}

File: legacy/dropzone47/session.py
import json
import os
import sqlite3
from typing import Any, Dict, Optional

from .config import SESSIONS_DB, logger

# In-memory session cache (persisted to SQLite for resilience across restarts).
user_sessions: Dict[int, Dict[str, Any]] = {}

# Track per-user active/finished downloads for listing and cancellation.
user_downloads: Dict[int, Dict[str, Any]] = {}

# SQLite file derived from the SESSIONS_DB base path.
_DB_PATH = f"{SESSIONS_DB}.sqlite3"


def _connect() -> sqlite3.Connection:
    parent = os.path.dirname(_DB_PATH)
    if parent:
        os.makedirs(parent, exist_ok=True)
    conn = sqlite3.connect(_DB_PATH)
    conn.execute(
        "CREATE TABLE IF NOT EXISTS sessions (user_id INTEGER PRIMARY KEY, data TEXT NOT NULL)"
    )
    return conn


def load_session(user_id: int) -> Optional[Dict[str, Any]]:
    try:
        with _connect() as conn:
            row = conn.execute(
                "SELECT data FROM sessions WHERE user_id = ?", (user_id,)
            ).fetchone()
        if row is not None:
            data: Dict[str, Any] = json.loads(row[0])
            return data
    except Exception as e:
        logger.warning("Failed to load session %s: %s", user_id, e)
    return None


def save_session(user_id: int, data: Dict[str, Any]) -> None:
    try:
        with _connect() as conn:
            conn.execute(
                "INSERT INTO sessions (user_id, data) VALUES (?, ?) "
                "ON CONFLICT(user_id) DO UPDATE SET data = excluded.data",
                (user_id, json.dumps(data)),
            )
    except Exception as e:
        logger.warning("Failed to save session %s: %s", user_id, e)


def delete_session(user_id: int) -> None:
    try:
        with _connect() as conn:
            conn.execute("DELETE FROM sessions WHERE user_id = ?", (user_id,))
    except Exception as e:
        logger.warning("Failed to delete session %s: %s", user_id, e)
